src/main/java/graphql/schema/DataFetchingFieldSelectionSetImpl.java

```diff
@@ -311,18 +316,28 @@ public String getFullyQualifiedName() {
         }
 
         @Override
-        public GraphQLObjectType getObjectType() {
-            return normalizedField.getObjectType();
+        public List<GraphQLFieldDefinition> getFieldDefinitions() {
+            return normalizedField.getFieldDefinitions(schema);
+        }
+
+        @Override
+        public GraphQLOutputType getType() {
+            return normalizedField.getType(schema);
+        }
+
+        @Override
+        public List<GraphQLObjectType> getObjectTypes() {
+            return (List) this.schema.getTypes(normalizedField.getObjectTypeNames());
         }
 
         @Override
-        public GraphQLFieldDefinition getFieldDefinition() {
-            return normalizedField.getFieldDefinition();
+        public List<String> getObjectTypeNames() {
+            return new ArrayList<>(normalizedField.getObjectTypeNames());
```

Comment:
ImmutableList.of() - no allocation if its immutable I think

Reply:
```
    public static <E> ImmutableList<E> copyOf(Collection<? extends E> elements) {
        if (elements instanceof ImmutableCollection) {
            ImmutableList<E> list = ((ImmutableCollection)elements).asList();
            return list.isPartialView() ? asImmutableList(list.toArray()) : list;
        } else {
            return construct(elements.toArray());
        }
    }

```

Yup no allocation on immutable collections

Reply:
well some but smarter

Reply:
unfortunately object types names is a mutable Set (on purpose)